classes/helper.php

```diff
@@ -81,47 +165,42 @@ public static function find_course_field(string $table): string {
-     * @param string $column The column to search.
+     * @param database_column_info $column The column to search.
      * @param int $limit The maximum number of results to return.
      * @return array The results of the search.
+     * @throws \dml_exception
      */
-    private static function plain_text_search(string $search, string $table,
-                                              string $column = self::ALL_COLUMNS, $limit = 0): array {
+    public static function plain_text_search(string $search, string $table,
+                                             database_column_info $column, int $limit = 0): array {
         global $DB;
 
         $results = [];
 
-        $columns = self::get_columns($table, $column);
-
         // Potential course field in the table.
         $coursefield = self::find_course_field($table);
 
-        // Table alias.
+        // Build query.
         $tablealias = 't';
+        $columnname = $DB->get_manager()->generator->getEncQuoted($column->name);
+        $searchsql = $DB->sql_like("$tablealias." . $columnname, '?', false);
+        $searchparam = '%'.$DB->sql_like_escape($search).'%';
+
+        if (!empty($coursefield)) {
+            $sql = "SELECT $tablealias.id,
+                           $tablealias.$columnname,
+                           $tablealias.$coursefield as courseid,
+                           c.idnumber as courseidnumber
+                      FROM {".$table."} t
+                 LEFT JOIN {course} c ON c.id = t.$coursefield
+                     WHERE $searchsql";
+        } else {
+            $sql = "SELECT id, $columnname FROM {".$table."} $tablealias WHERE $searchsql";
+        }
 
-        foreach ($columns as $column) {
-            $columnname = $DB->get_manager()->generator->getEncQuoted($column->name);
-
-            $searchsql = $DB->sql_like("$tablealias." . $columnname, '?', false);
-            $searchparam = '%'.$DB->sql_like_escape($search).'%';
-
-            if (!empty($coursefield)) {
-                $sql = "SELECT $tablealias.id,
-                               $tablealias.$columnname,
-                               $tablealias.$coursefield as courseid,
-                               c.idnumber as courseidnumber
-                          FROM {".$table."} t
-                     LEFT JOIN {course} c ON c.id = t.$coursefield
-                         WHERE $searchsql";
-            } else {
-                $sql = "SELECT id, $columnname FROM {".$table."} $tablealias WHERE $searchsql";
-            }
-
-            if ($column->meta_type === 'X' || $column->meta_type === 'C') {
-                $records = $DB->get_records_sql($sql, [$searchparam], 0, $limit);
-                if ($records) {
-                    $results[$table][$column->name] = $records;
-                }
+        if ($column->meta_type === 'X' || $column->meta_type === 'C') {
+            $records = $DB->get_records_sql($sql, [$searchparam], 0, $limit);
```

Comment:
we need to convert these to get_recordset_sql calls and then stream the results to the file handle

So this function will not return results, just have side effects. It can return the row count it found instead

Reply:
Make change but a bit different  to your thought:

If stream is specify -> output to file
otherwise return the record set to result: this will be refactor later when we need to use  it in webui